subsys/net/conn_mgr/online_checker.c

```diff
@@ -0,0 +1,935 @@
+		ret = do_online_http_check(iface, online_check->url);
+	} else {
+		if (online_check->strategy != NET_CONN_MGR_ONLINE_CHECK_PING &&
+		    online_check->strategy != NET_CONN_MGR_ONLINE_CHECK_HTTP) {
+			NET_ERR("Invalid online check strategy (%d)", online_check->strategy);
+		}
+
+		return -EINVAL;
+	}
+
+	if (ret == 0) {
+		online_check->running = true;
+	} else {
+		online_check->running = false;
+	}
+
+	return ret;
+}
+
+static void clear_online_check_flags(void)
+{
+	int i, state_count;
+
+	k_mutex_lock(&conn_mgr_mon_lock, K_FOREVER);
+
+	state_count = conn_mgr_get_iface_states(&states);
+
+	for (i = 0; i < state_count; i++) {
+		states[i] &= ~CONN_MGR_IF_ONLINE_CHECK;
+	}
+
+	k_mutex_unlock(&conn_mgr_mon_lock);
+}
+
+static void stop_online_check(void)
+{
+	int ret;
+
+	clear_online_check_flags();
+
+	k_mutex_lock(&check_running_lock, K_FOREVER);
+	check_running = false;
+	k_mutex_unlock(&check_running_lock);
+
+	if (!IS_ENABLED(CONFIG_NET_CONNECTION_MANAGER_ONLINE_VERIFY)) {
+		return;
+	}
+
+	if (net_trickle_is_running(&online_check->verifier)) {
+		ret = net_trickle_stop(&online_check->verifier);
+		if (ret < 0) {
+			NET_DBG("Cannot stop Trickle timer (%d)", ret);
+		}
+
+		k_work_cancel_delayable(&online_check->verifier_work);
+	}
+
+	/* If there is a check currently ongoing, it will eventually stop
+	 * and the new check will be done using the new strategy.
+	 */
+}
+
+void conn_mgr_online_connectivity_check(void)
+{
+	int ret, i, state_count;
+
+	k_mutex_lock(&conn_mgr_mon_lock, K_FOREVER);
+	state_count = conn_mgr_get_iface_states(&states);
+	k_mutex_unlock(&conn_mgr_mon_lock);
+
+	for (i = 0; i < state_count; i++) {
+		if (states[i] & CONN_MGR_IF_READY) {
+			k_mutex_lock(&conn_mgr_mon_lock, K_FOREVER);
+
+			if (!(states[i] & CONN_MGR_IF_READY) ||
+			    (states[i] & CONN_MGR_IF_ONLINE_CHECK)) {
+				k_mutex_unlock(&conn_mgr_mon_lock);
+				continue;
+			}
+
+			k_mutex_unlock(&conn_mgr_mon_lock);
+
+			/* Do connectivity check and if it succeeds,
+			 * generate CONNECTIVITY_ONLINE event. The online
+			 * check might take some time.
+			 */
+			k_mutex_lock(&check_running_lock, K_FOREVER);
+
+			if (check_running) {
+				NET_DBG("Online check already running");
+				k_mutex_unlock(&check_running_lock);
+				break;
+			}
+
+			check_running = true;
+
+			k_mutex_unlock(&check_running_lock);
+
+			k_mutex_lock(&conn_mgr_mon_lock, K_FOREVER);
+
+			/* Tells that we are doing online check for this interface */
+			states[i] |= CONN_MGR_IF_ONLINE_CHECK;
+
+			k_mutex_unlock(&conn_mgr_mon_lock);
+
+			/* Did we get to online? This is a synchronous call and
+			 * in worst case might take a long time to return.
+			 */
+			ret = do_online_check(conn_mgr_mon_get_if_by_index(i));
+			if (ret == 0) {
+				/* We are online, no need to check the other interfaces */
+				break;
+			}
+
+			k_mutex_lock(&conn_mgr_mon_lock, K_FOREVER);
+
+			states[i] &= ~CONN_MGR_IF_ONLINE_CHECK;
```

Comment:
You are very careful to only ever have a single iface with the `CONN_MGR_IF_ONLINE_CHECK` flag set true at any time. (you always clear this flag here before moving on to checking the next iface).

So, why can't this just be a pointer to or index representing the iface currently marked online? Set it to -1/NULL when none are marked as online.

The only reason that I can see to use flags is if more than one iface can be marked as `CONN_MGR_IF_ONLINE_CHECK` at once, but it doesn't seem that is true?

Earlier you said:
>  I think we need it in order to know which interfaces were done online check

But if only one of these ifaces ever has `CONN_MGR_IF_ONLINE_CHECK` set at a time, I don't think there is an issue with using just a pointer or index.

I know that earlier I pushed for ***more*** flags, but at that time I severely misunderstood your goals with these additions. Now that I understand the goal, I think per-iface flags are not necessary at all